toonz/sources/toonzlib/preferences.cpp

```diff
@@ -563,675 +319,491 @@ Preferences::Preferences()
+      value.setValue(size);
+    }
+    break;
+  case QMetaType::QMetaType::QColor:
+    if (m_settings->contains(idString)) {
+      QString str = m_settings->value(idString).toString();
+      value.setValue(stringToColor(str));
+    }
+    // following two conditions are to keep compatibility with older versions
+    else if (m_settings->contains(idString + "_R")) {
+      QColor color = value.value<QColor>();
+      color.setRed(m_settings->value(idString + "_R", color.red()).toInt());
+      color.setGreen(m_settings->value(idString + "_G", color.green()).toInt());
+      color.setBlue(m_settings->value(idString + "_B", color.blue()).toInt());
+      color.setAlpha(m_settings->value(idString + "_M", color.alpha()).toInt());
+      value.setValue(color);
+    } else if (m_settings->contains(idString + ".r")) {
+      QColor color = value.value<QColor>();
+      color.setRed(m_settings->value(idString + ".r", color.red()).toInt());
+      color.setGreen(m_settings->value(idString + ".g", color.green()).toInt());
+      color.setBlue(m_settings->value(idString + ".b", color.blue()).toInt());
+      color.setAlpha(255);
+      value.setValue(color);
+    }
+    break;
+  case QMetaType::QVariantMap:  // used in colorCalibrationLutPaths
+    if (m_settings->contains(idString) &&
+        m_settings->value(idString).canConvert(type)) {
+      QMap<QString, QString> pathMap;
+      QAssociativeIterable iterable =
+          m_settings->value(idString).value<QAssociativeIterable>();
+      QAssociativeIterable::const_iterator it        = iterable.begin();
+      const QAssociativeIterable::const_iterator end = iterable.end();
+      for (; it != end; ++it)
+        pathMap.insert(it.key().toString(), it.value().toString());
+      value.setValue(pathMap);
+    }
+    break;
+  default:
+    std::cout << "Unsupported type detected" << std::endl;
+    // load anyway
+    value = m_settings->value(idString, value);
+    break;
+  }
 
-void Preferences::enableRewindAfterPlayback(bool on) {
-  m_rewindAfterPlaybackEnabled = on;
-  m_settings->setValue("rewindAfterPlayback", on ? "1" : "0");
+  m_items.insert(id, PreferencesItem(idString, type, value, min, max));
 }
 
 //-----------------------------------------------------------------
 
-void Preferences::enableFitToFlipbook(bool on) {
-  m_fitToFlipbookEnabled = on;
-  m_settings->setValue("fitToFlipbook", on ? "1" : "0");
+void Preferences::setCallBack(const PreferencesItemId id, OnEditedFunc func) {
+  getItem(id).onEditedFunc = func;
 }
 
 //-----------------------------------------------------------------
 
-void Preferences::enablePreviewAlwaysOpenNewFlip(bool on) {
-  m_previewAlwaysOpenNewFlipEnabled = on;
-  m_settings->setValue("previewAlwaysOpenNewFlip", on ? "1" : "0");
+PreferencesItem &Preferences::getItem(const PreferencesItemId id) {
+  assert(m_items.contains(id));
+  if (!m_items.contains(id)) return PreferencesItem();
```

Comment:
Linux and macOS builds are failing here with this message

/Users/travis/build/opentoonz/opentoonz/toonz/sources/toonzlib/preferences.cpp:668:37: error: non-const lvalue reference to type 'PreferencesItem' cannot bind to a temporary of type 'PreferencesItem'
  if (!m_items.contains(id)) return PreferencesItem();

Reply:
Thank you for the notice, I've made the same mistake again.
I removed the problematic line.